Add TYNDP data processing & link generation

brynpickering opened the pull request:
Fixes #61.

As discussed @timtroendle, this has not been added as an override, but rather a 'hanging' file. That is, it will be generated but not pointed to by `national/model.yaml`. A user has to update `model.yaml` to enforce these links.

## Checklist

Any checks which are not relevant to the PR can be pre-checked by the PR creator. All others should be checked by the reviewer. You can add extra checklist items here if required by the PR.

- [ ] CHANGELOG updated
- [ ] Minimal workflow tests pass
- [ ] Tests added to cover contribution
- [ ] Documentation updated
- [ ] Configuration schema updated

timtroendle reviewed (changes requested):
Very nice! Some minor comments below.

One major comment: there should be tests. There should be an integration test testing that the model runs and Gurobi finds an optimum (very easy to implement, see my comment) but there should also be a few tests testing for actual capacity constraint numbers. Those tests could be on the integration test level (i.e. the Calliope model object)  or on the unit-test level. I'd prefer the latter (or both).

brynpickering commented on Snakefile:
```diff
@@ -25,7 +25,7 @@ include: "./rules/shapes.smk"
 include: "./rules/sync.smk"
 localrules: all, download_raw_load, model, clean, parameterise_template, download_potentials, download_eurocalliope_dataset
-localrules: download_capacity_factors_wind_and_solar
+localrules: download_capacity_factors_wind_and_solar, download_enstoe_tyndp_zip
```
Agreed. As in #145 for YAMLs, this should probably follow some clear modularisation effort.

brynpickering commented on scripts/link_entsoe_tyndp.py:
```diff
@@ -0,0 +1,99 @@
+    average_ntc = getattr(pd.concat([tyndp_import, tyndp_export]).abs(), entsoe_ntc_limit)(level=["loc_from", "loc_to"])
+
+    # Only keep countries found in the model
+    return average_ntc[
```
Sadly, `xs` doesn't work with multiple values per level accessed. But I can probably use the pandas `IndexSlice`.

brynpickering commented on scripts/link_entsoe_tyndp.py:
```diff
@@ -0,0 +1,99 @@
+
+
```
How much is the fix offset by the energy consumption of your comment?

timtroendle replied:
0.0000000000001 kg CO₂. The break-even will be in 10 years time. But we _really_ need to end this conversation as we are moving it more and more into the future.

brynpickering commented:
@timtroendle I've addressed your issues, but now the minimal workflow is failing due to 'electrical islands' in the generation of `build/model/national/link-all-neighbours.yaml`. I did change the pointer to the CRS in this file, but having tested it with reverting the CRS string, the same issue exists. Any ideas on why this might be happening?

timtroendle commented on Snakefile:
```diff
@@ -275,6 +275,43 @@ rule electricity_load:
+        scaling_factor = config["scaling-factors"]["power"]
+    output: "build/model/{resolution}/entsoe-tyndp-links.yaml"
+    wildcard_constraints:
+        resolution = "national"
```
Why do you make it a wildcard if there is anyway just a single value? Why not 

`output: "build/model/national/entsoe-tyndp-links.yaml"`?

brynpickering replied:
Because snakemake doesn't like it. I can update the pointer from the `model` rule to be to the filepath rather than the rule output, which would solve this. However, it seems good practice to point to the rule output rather than the filepath... Do you prefer I change to the filepath in the `model` rule?

timtroendle replied:
Ah, I see. Well I'd change from rule to filepath and remove the wildcard. But the current way works too of course.

timtroendle commented on templates/README.md:
```diff
@@ -66,6 +66,14 @@ In euro-calliope, we model load shedding not as actual reduction of demand but a
 
+## Manipulating the model using file imports
+
+The `example-model.yaml` configuration file in each resolution sub-directory (e.g. `national/model.yaml`) specifies a list of other configuration files to bring together to describe the model. This list can be changed by the modeller to select a combination of different configuration files. The final result is similar to the use of overrides, except that that model is *never* aware of the configu ...
```
`national/model.yaml` -> `national/example-model.yaml`

timtroendle commented on tests/resources/national/model.yaml:
```diff
@@ -15,16 +15,20 @@ model:
     shed-load: ["load-shedding"]
 overrides:
-    connect:
+    connect_all_neighbours:
```
Ok. For consistency, can you rename it in `tests/resources/regional/model.yaml`, too?

timtroendle commented on templates/README.md:
```diff
@@ -66,6 +66,14 @@ In euro-calliope, we model load shedding not as actual reduction of demand but a
+
+> Transmission links
+
+There are no transmission links specified in `example-model.yaml` by default. At all resolutions, `link-all-neighbours.yaml` can be added to link neighbouring regions + a selection of pre-defined sub-sea links, but with no capacity limits. At the national resolution, transmission links can be set based on an ENTSO-E ten-year development plan 2020 scenario (`national/entsoe-tyndp-links.yaml`). The ...
```
Hmm, that's not correct, is it? The example model imports `link-all-neighbours.yaml` and therefore, by default, all neighbours are connected.

timtroendle commented on tests/scripts/test_link_entsoe_tyndp.py:
```diff
@@ -0,0 +1,25 @@
+@pytest.fixture
+def sliced_df():
+    def _sliced_df(idx):
+        data = list(range(10, 10 + len(idx)))
```
You don't need to convert this to a list here, you should be able to hand over data as an iterator.

timtroendle commented on tests/scripts/test_link_entsoe_tyndp.py:
```diff
@@ -0,0 +1,25 @@
+    input_df = sliced_df(input_idx)
+    cleaned_series = _split_links_in_index(input_df)
+    assert cleaned_series.index.symmetric_difference(expected_idx).empty
+    assert np.allclose(cleaned_series.loc[expected_idx].values, expected_values)
```
I thought about this for a while, but I still cannot figure out what you are testing in this second assert statement. I don't get what the expected values are and why.

brynpickering replied:
Let me know if the latest update is clearer

timtroendle replied:
It is!

timtroendle reviewed (approved):
Good!

I think there is a mistake in the documentation that needs to be fixed, but apart from that looks good.
